### ISSUE TITLE
Cannot run AppImages with sudo (on Debian 10)

### ISSUE THREAD

**Issue by colemanlee** [TheAssassin/AppImageLauncher#266]
Hi,
I also meet this problem when I run `./pling-store-5.0.1-1-x86_64.AppImage` in ROOT graphical interface.
* My system information:
```
Linux debian 5.3.0-1-686-pae #1 SMP Debian 5.3.7-1 (2019-10-19) i686 GNU/Linux
```
* deb package information：
```
# apt-cache show appimagelauncher
Package: appimagelauncher
Status: install ok installed
Priority: optional
Section: devel
Installed-Size: 56454
Maintainer: TheAssassin
Architecture: i386
Version: 2.0.2-travis883~f5ac870+buster
Description: AppImageLauncher built using CMake
Description-md5: 604b265bc489c2475cd6e25aabba693d
```
![PIC__28](https://user-images.githubusercontent.com/10258421/68522719-7d3ce600-02e9-11ea-9483-e2ea6e416915.jpg)

Yes, I modified the system and log in the system graphical interface with `root` for some test purpose.
Then I tried to login normal user mode in the graphical interface, and run the command:
```
sudo ./HastyHeroes-1.0.0-x86_64.AppImage 
```
The error appear:
```
[sudo] jeff 's password：
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
Failed to connect to bus: no files or dir
Failed to connect to bus: no files or dir
---->dialog box "Desktop Intergration—AppImageLauncher" appear.
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 565, resource id: 10653673, major code: 40 (TranslateCoords), minor code: 0
---->Click "Run Once" button
Failed to connect to bus: no files or dir
Failed to connect to bus: no files or dir
---->dialog box "Error—AppImageLauncher" appear and click "OK" button.
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 699, resource id: 10654207, major code: 40 (TranslateCoords), minor code: 0
```
![PIC__33](https://user-images.githubusercontent.com/10258421/68528706-50142600-0331-11ea-9f44-1fade9f38ca4.jpg)
![PIC__34](https://user-images.githubusercontent.com/10258421/68528707-51455300-0331-11ea-956a-801266c14fd1.jpg)

If I run the command:
```
./HastyHeroes-1.0.0-x86_64.AppImage
```
The error appears when I click the "Run once" button:
```
execv() failed: executable file format error
```
![PIC__35](https://user-images.githubusercontent.com/10258421/68528730-9ec1c000-0331-11ea-8161-971e58e13fc4.jpg)

I have no idea how to run AppImage on Debian 10 now.
I open a new issue in order to allow for debugging.

**Comment by TheAssassin** [comment]
Running as `root` user is a *really* nonstandard (since pretty unsafe) situation. Debian is also pretty annoying about how `sudo` works. I guess AppImageLauncher gets confused because I made a few too many assumptions about how users would use the software. (I don't want to say Debian is annoying, but... well... sometimes, they *really* live in the past...)

There are two use cases to regard: using AppImageLauncher through `sudo` (especially relevant on Debian, where `sudo` works differently than on most other systems), and using AppImageLauncher when logged in as `root` directly.

It seems like in general the UI works as intended through sudo. So far, so good, the warnings can be ignored IMO. Logged in as `root`, I wouldn't expect any issues anyway.

Something I *never* tested was running AppImages when being logged in as `root`, I have to admit. I do see how this causes issues. My approach would be to completely "disable" AppImageLauncher for `root` (in the sense that it wouldn't show any dialogs). I don't think system administrators are interested in the features it provides. However, it still requires AppImageLauncherFS. Unfortunately, there's nothing like a "systemd user service" for root users, as far as I can see.

I think I already know where the problem in `sudo` mode is. To check if the solution I think of works, can you please run the following snippet on your computer and provide the output? (Please note the special quoting, it'll be best if you just copy and paste the entire command)

```sh
sudo bash -c 'echo "USER: $USER -- SUDO_USER: $SUDO_USER"'
# should yield: USER: root -- SUDO_USER: jeff
# if it does, the fix will be relatively easy
```

Let's make "login as `root`" a separate issue, and deal with "running through `sudo` (on Debian)" in this one.

**Comment by TheAssassin** [comment]
TL;DR: this is *not* a general Debian 10 issue, it's specific to running AppImages through `sudo` when AIL is installed.

**Comment by colemanlee** [comment]
> Let's make "login as `root`" a separate issue, and deal with "running through `sudo` (on Debian)" in this one.

Yes,
```
jeff@debian:~$ sudo bash -c 'echo "USER: $USER -- SUDO_USER: $SUDO_USER"'
USER: root -- SUDO_USER: jeff
```

**Comment by RogueScholar** [comment]
Just to add my experience here while I'm here reading the thread, I've also encountered this issue with AIL, and I think it was through a legitimate use case no less. It's something I only do at barest necessity, namely creating a Windows 10 installation flash drive, but on those rare occasions there's no tool as reliable and robust for that task in Linux as the Windows2USB AppImage (trust me, I've tried them all).

Naturally it's something that needs to be invoked with elevated privileges and in the dark ages before I found AIL that always went off without a hitch. The most recent time I needed it though, I ran head-first into the same brick wall as @colemanlee. I tried everything I could think of to work around the issue, including setting the XDG_RUNTIME_DIR and DBUS_SESSION_BUS_ADDRESS environment variables and some of the more obscure systemd key service file keys to loosen the restrictions on the process, but in the end was completely stymied.

Rather than uninstall AIL even temporarily though, I resorted to mounting the AppImage and stealing its contents wholesale to place in /usr/local and accomplishing things that way. In the end I chalked it up to a reasonable constraint on functionality, mostly because I consider myself fairly handy with systemd at this point and since I couldn't identify a valid configuration to produce a better result it felt in poor taste to dump the issue in someone else's lap. I'm only mentioning it here to establish that there are a handful of really good AppImages in circulation that _only_ work when called via sudo or within an elevated shell, and if a way were found to make it so that using them and using AIL wasn't an either/or state like it is now, I think that'd be ideal.
